File: grind-75/tree/diameterOfBT.py
from turtle import reset
from typing import Optional

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        res = 0

        def dfs(root):
            nonlocal res
            if not root:
                return 0
            l = dfs(root.left)
            r = dfs(root.right)
            res = max(res, l + r)
            return max(l, r) + 1

        dfs(root)
        return res

File: projects/grind-75/typescript/tree/diameter-of-bt.ts
import { TreeNode } from './common';

export default function diameterOfBinaryTree(root: TreeNode | null): number {
  let res = 0;

  function dfs(node: TreeNode | null = root): number {
    if (!node) {
      return 0;
    }

    const left = dfs(node.left);
    const right = dfs(node.right);
    res = Math.max(res, left + right);
    return Math.max(left, right) + 1;
  }

  dfs();
  return res;
}
